<!DOCTYPE html>
<html lang="zh-CN">
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="robots" content="noodp" />
        <meta http-equiv="X-UA-Compatible" content="IE=edge, chrome=1">
        <title>不同scRNAseq批次效应矫正方法计算比较 - 美吉生物信息团队</title><meta name="Description" content="不同scRNAseq批次效应矫正方法"><meta property="og:title" content="不同scRNAseq批次效应矫正方法计算比较" />
<meta property="og:description" content="不同scRNAseq批次效应矫正方法" />
<meta property="og:type" content="article" />
<meta property="og:url" content="http://czheluo.github.io/MajorbioHuGo/posts/scrnaseq-batch-correction/" />
<meta property="og:image" content="http://czheluo.github.io/MajorbioHuGo/posts/scrnaseq-batch-correction/batch.png"/>
<meta property="article:published_time" content="2020-12-14T16:29:59+08:00" />
<meta property="article:modified_time" content="2020-12-15T16:29:59+08:00" />
<meta name="twitter:card" content="summary_large_image"/>
<meta name="twitter:image" content="http://czheluo.github.io/MajorbioHuGo/posts/scrnaseq-batch-correction/batch.png"/>
<meta name="twitter:title" content="不同scRNAseq批次效应矫正方法计算比较"/>
<meta name="twitter:description" content="不同scRNAseq批次效应矫正方法"/>
<meta name="application-name" content="美吉生物信息团队">
<meta name="apple-mobile-web-app-title" content="美吉生物信息团队"><link rel="shortcut icon" type="image/x-icon" href="/favicon.ico" />
        <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
        <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png"><link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png"><link rel="manifest" href="/site.webmanifest"><link rel="canonical" href="http://czheluo.github.io/MajorbioHuGo/posts/scrnaseq-batch-correction/" /><link rel="prev" href="http://czheluo.github.io/MajorbioHuGo/posts/mapping-make-fake-data/" /><link rel="next" href="http://czheluo.github.io/MajorbioHuGo/posts/probabilistic-estimation-of-expression-residuals/" /><link rel="stylesheet" href="/MajorbioHuGo/lib/normalize/normalize.min.css"><link rel="stylesheet" href="/MajorbioHuGo/css/style.min.css"><link rel="stylesheet" href="/MajorbioHuGo/lib/fontawesome-free/all.min.css"><link rel="stylesheet" href="/MajorbioHuGo/lib/animate/animate.min.css"><script type="application/ld+json">
    {
        "@context": "http://schema.org",
        "@type": "BlogPosting",
        "headline": "不同scRNAseq批次效应矫正方法计算比较",
        "inLanguage": "zh-CN",
        "mainEntityOfPage": {
            "@type": "WebPage",
            "@id": "http:\/\/czheluo.github.io\/MajorbioHuGo\/posts\/scrnaseq-batch-correction\/"
        },"image": [{
                            "@type": "ImageObject",
                            "url": "http:\/\/czheluo.github.io\/MajorbioHuGo\/posts\/scrnaseq-batch-correction\/batch.png",
                            "width":  1198 ,
                            "height":  621 
                        }],"genre": "posts","keywords": "批次效应, scRNAseq, RNAseq, 单细胞","wordcount":  743 ,
        "url": "http:\/\/czheluo.github.io\/MajorbioHuGo\/posts\/scrnaseq-batch-correction\/","datePublished": "2020-12-14T16:29:59+08:00","dateModified": "2020-12-15T16:29:59+08:00","publisher": {
            "@type": "Organization",
            "name": "Xiaolong He"},"author": {
                "@type": "Person",
                "name": "Xiaolong He"
            },"description": "不同scRNAseq批次效应矫正方法"
    }
    </script></head>
    <body header-desktop="" header-mobile=""><script type="text/javascript">(window.localStorage && localStorage.getItem('theme') ? localStorage.getItem('theme') === 'dark' : ('' === 'auto' ? window.matchMedia('(prefers-color-scheme: dark)').matches : '' === 'dark')) && document.body.setAttribute('theme', 'dark');</script>

        <div id="mask"></div><div class="wrapper"><header class="desktop" id="header-desktop">
    <div class="header-wrapper">
        <div class="header-title">
            <a href="/MajorbioHuGo/" title="美吉生物信息团队">美吉生物信息团队</a>
        </div>
        <div class="menu">
            <div class="menu-inner"><a class="menu-item" href="/MajorbioHuGo/MajorbioHuGo/posts/"> 所有文章 </a><a class="menu-item" href="/MajorbioHuGo/MajorbioHuGo/tags/"> 标签 </a><a class="menu-item" href="/MajorbioHuGo/MajorbioHuGo/categories/"> 分类 </a><a class="menu-item" href="/MajorbioHuGo/MajorbioHuGo/categories/documentation/"> 文档 </a><a class="menu-item" href="/MajorbioHuGo/MajorbioHuGo/about/"> 关于 </a><a class="menu-item" href="/MajorbioHuGo" title="GitHub"><i class='fab fa-github fa-fw'></i>  </a><span class="menu-item delimiter"></span><span class="menu-item search" id="search-desktop">
                        <input type="text" placeholder="搜索文章标题或内容..." id="search-input-desktop">
                        <a href="javascript:void(0);" class="search-button search-toggle" id="search-toggle-desktop" title="搜索">
                            <i class="fas fa-search fa-fw"></i>
                        </a>
                        <a href="javascript:void(0);" class="search-button search-clear" id="search-clear-desktop" title="清空">
                            <i class="fas fa-times-circle fa-fw"></i>
                        </a>
                        <span class="search-button search-loading" id="search-loading-desktop">
                            <i class="fas fa-spinner fa-fw fa-spin"></i>
                        </span>
                    </span><a href="javascript:void(0);" class="menu-item theme-switch" title="切换主题">
                    <i class="fas fa-adjust fa-fw"></i>
                </a>
            </div>
        </div>
    </div>
</header><header class="mobile" id="header-mobile">
    <div class="header-container">
        <div class="header-wrapper">
            <div class="header-title">
                <a href="/MajorbioHuGo/" title="美吉生物信息团队">美吉生物信息团队</a>
            </div>
            <div class="menu-toggle" id="menu-toggle-mobile">
                <span></span><span></span><span></span>
            </div>
        </div>
        <div class="menu" id="menu-mobile"><div class="search-wrapper">
                    <div class="search mobile" id="search-mobile">
                        <input type="text" placeholder="搜索文章标题或内容..." id="search-input-mobile">
                        <a href="javascript:void(0);" class="search-button search-toggle" id="search-toggle-mobile" title="搜索">
                            <i class="fas fa-search fa-fw"></i>
                        </a>
                        <a href="javascript:void(0);" class="search-button search-clear" id="search-clear-mobile" title="清空">
                            <i class="fas fa-times-circle fa-fw"></i>
                        </a>
                        <span class="search-button search-loading" id="search-loading-mobile">
                            <i class="fas fa-spinner fa-fw fa-spin"></i>
                        </span>
                    </div>
                    <a href="javascript:void(0);" class="search-cancel" id="search-cancel-mobile">
                        取消
                    </a>
                </div><a class="menu-item" href="/MajorbioHuGo/MajorbioHuGo/posts/" title="">所有文章</a><a class="menu-item" href="/MajorbioHuGo/MajorbioHuGo/tags/" title="">标签</a><a class="menu-item" href="/MajorbioHuGo/MajorbioHuGo/categories/" title="">分类</a><a class="menu-item" href="/MajorbioHuGo/MajorbioHuGo/categories/documentation/" title="">文档</a><a class="menu-item" href="/MajorbioHuGo/MajorbioHuGo/about/" title="">关于</a><a class="menu-item" href="/MajorbioHuGo" title="GitHub"><i class='fab fa-github fa-fw'></i></a><a href="javascript:void(0);" class="menu-item theme-switch" title="切换主题">
                <i class="fas fa-adjust fa-fw"></i>
            </a></div>
    </div>
</header>
<div class="search-dropdown desktop">
    <div id="search-dropdown-desktop"></div>
</div>
<div class="search-dropdown mobile">
    <div id="search-dropdown-mobile"></div>
</div>
<main class="main">
                <div class="container"><div class="toc" id="toc-auto">
            <h2 class="toc-title">目录</h2>
            <div class="toc-content" id="toc-content-auto"></div>
        </div><article class="page single"><h1 class="single-title animated flipInX">不同scRNAseq批次效应矫正方法计算比较</h1><div class="post-meta">
            <div class="post-meta-line"><span class="post-author"><a href="https://github.com/xiaolong-he999" title="Author" target="_blank" rel="noopener noreffer author" class="author"><i class="fas fa-user-circle fa-fw"></i>Xiaolong He</a></span>&nbsp;<span class="post-category">收录于 <a href="/MajorbioHuGo/categories/tools/"><i class="far fa-folder fa-fw"></i>tools</a></span></div>
            <div class="post-meta-line"><i class="far fa-calendar-alt fa-fw"></i>&nbsp;<time datetime="2020-12-14">2020-12-14</time>&nbsp;<i class="fas fa-pencil-alt fa-fw"></i>&nbsp;约 743 字&nbsp;
                <i class="far fa-clock fa-fw"></i>&nbsp;预计阅读 2 分钟&nbsp;</div>
        </div><div class="featured-image"><img
        class="lazyload"
        src="/MajorbioHuGo/svg/loading.min.svg"
        data-src="/MajorbioHuGo/posts/scrnaseq-batch-correction/batch.png"
        data-srcset="/MajorbioHuGo/posts/scrnaseq-batch-correction/batch.png, /MajorbioHuGo/posts/scrnaseq-batch-correction/batch.png 1.5x, /MajorbioHuGo/posts/scrnaseq-batch-correction/batch.png 2x"
        data-sizes="auto"
        alt="/MajorbioHuGo/posts/scrnaseq-batch-correction/batch.png"
        title="不同scRNAseq批次效应矫正方法" /></div><div class="details toc" id="toc-static"  kept="">
                <div class="details-summary toc-title">
                    <span>目录</span>
                    <span><i class="details-icon fas fa-angle-right"></i></span>
                </div>
                <div class="details-content toc-content" id="toc-content-static"><nav id="TableOfContents">
  <ul>
    <li><a href="#limma">limma</a></li>
    <li><a href="#harmony">harmony</a></li>
    <li><a href="#liger">LIGER</a></li>
    <li><a href="#mnn-correct">MNN Correct</a></li>
    <li><a href="#fastmnn">fastMNN</a></li>
  </ul>
</nav></div>
            </div><div class="content" id="content"><h1 id="batch-methds">Batch methds</h1>
<p>所用数据:</p>
<pre><code>/mnt/ilustre/centos7users/xiaolong.he/project/scRNA/new_pipe/04.Step4.combain/Merged.Seurat.rds
</code></pre><h2 id="limma">limma</h2>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="c1">#!/usr/bin/env Rscript3.6.1</span>
<span class="nf">suppressMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;getopt&#34;</span><span class="p">))</span>
<span class="nf">options</span><span class="p">(</span><span class="n">bitmapType</span><span class="o">=</span><span class="s">&#39;cairo&#39;</span><span class="p">)</span>
<span class="n">spec</span> <span class="o">=</span> <span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span>
    <span class="s">&#39;rds&#39;</span><span class="p">,</span><span class="s">&#39;r&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
    <span class="s">&#39;outdir&#39;</span><span class="p">,</span><span class="s">&#39;o&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
    <span class="s">&#39;help&#39;</span><span class="p">,</span><span class="s">&#39;h&#39;</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="s">&#39;logical&#39;</span>
    <span class="p">),</span> <span class="n">byrow</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="m">4</span><span class="p">);</span>
<span class="n">opt</span> <span class="o">=</span> <span class="nf">getopt</span><span class="p">(</span><span class="n">spec</span><span class="p">);</span>
<span class="n">print_usage</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">spec</span><span class="o">=</span><span class="kc">NULL</span><span class="p">){</span>
  <span class="nf">cat</span><span class="p">(</span><span class="nf">getopt</span><span class="p">(</span><span class="n">spec</span><span class="p">,</span> <span class="n">usage</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">));</span>
  <span class="nf">cat</span><span class="p">(</span><span class="s">&#34;Usage example: \n&#34;</span><span class="p">);</span>
  <span class="nf">cat</span><span class="p">(</span><span class="s">&#34;Usage example:
</span><span class="s">    Rscript limma.r --i  --o
</span><span class="s">    Usage:
</span><span class="s">    --rds       input file rds
</span><span class="s">    --outdir    outdir
</span><span class="s">    --help              usage\n&#34;</span>
    <span class="p">);</span>
  <span class="nf">q</span><span class="p">(</span><span class="n">status</span><span class="o">=</span><span class="m">1</span><span class="p">);</span>
<span class="p">}</span>

<span class="nf">if </span><span class="p">(</span> <span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">help</span><span class="p">))</span> <span class="p">{</span><span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">rds</span><span class="p">)){</span><span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">)){</span><span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)}</span>


<span class="nf">suppressMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;Seurat&#34;</span><span class="p">))</span>

<span class="n">scRNA</span> <span class="o">&lt;-</span> <span class="nf">readRDS</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">rds</span><span class="p">)</span>
<span class="n">count</span> <span class="o">&lt;-</span> <span class="nf">as.matrix</span><span class="p">(</span><span class="n">scRNA</span><span class="o">@</span><span class="n">assays</span><span class="o">$</span><span class="n">RNA</span><span class="o">@</span><span class="n">counts</span><span class="p">)</span>
<span class="n">count</span> <span class="o">&lt;-</span> <span class="nf">log</span><span class="p">(</span><span class="n">count</span><span class="m">+1</span><span class="p">)</span>
<span class="n">batchs</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s">&#39;1&#39;</span><span class="p">,</span><span class="m">7205</span><span class="p">),</span><span class="nf">rep</span><span class="p">(</span><span class="s">&#39;2&#39;</span><span class="p">,</span><span class="m">9104</span><span class="p">))</span>
<span class="n">lim_batch</span> <span class="o">&lt;-</span> <span class="n">limma</span><span class="o">::</span><span class="nf">removeBatchEffect</span><span class="p">(</span><span class="n">count</span><span class="p">,</span><span class="n">batch</span><span class="o">=</span><span class="n">batchs</span><span class="p">)</span>
<span class="nf">write.table</span><span class="p">(</span><span class="n">lim_batch</span><span class="p">,</span><span class="n">file</span><span class="o">=</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">,</span><span class="s">&#34;limma.txt&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#34;&#34;</span><span class="p">),</span><span class="n">col.names</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span><span class="n">row.names</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span><span class="n">quote</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="nf">save.image</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">,</span><span class="s">&#34;limma.RData&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#34;&#34;</span><span class="p">))</span>
<span class="nf">proc.time</span><span class="p">()</span>
</code></pre></div><p>limma去批次后矩阵含有大量负值，针对大量负值无法进行差异表达基因计算。
可通过对负值赋值为0后计算。</p>
<p><img
        class="lazyload"
        src="/MajorbioHuGo/svg/loading.min.svg"
        data-src="liger_DE_num.png"
        data-srcset="/MajorbioHuGo/posts/scrnaseq-batch-correction/liger_DE_num.png, liger_DE_num.png 1.5x, /MajorbioHuGo/posts/scrnaseq-batch-correction/liger_DE_num.png 2x"
        data-sizes="auto"
        alt="/MajorbioHuGo/posts/scrnaseq-batch-correction/liger_DE_num.png"
        title="liger_DE_num.png" />
<img
        class="lazyload"
        src="/MajorbioHuGo/svg/loading.min.svg"
        data-src="limma_DE.png"
        data-srcset="/MajorbioHuGo/posts/scrnaseq-batch-correction/limma_DE.png, limma_DE.png 1.5x, /MajorbioHuGo/posts/scrnaseq-batch-correction/limma_DE.png 2x"
        data-sizes="auto"
        alt="/MajorbioHuGo/posts/scrnaseq-batch-correction/limma_DE.png"
        title="limma_DE.png" /></p>
<p>参考文献：Ritchie ME, Phipson B, Wu D, Hu Y, Law CW, Shi W and Smyth GK (2015). <a href="https://academic.oup.com/nar/article/43/7/e47/2414268" target="_blank" rel="noopener noreffer">limma powers differential expression analyses for RNA-sequencing and microarray studies</a>. Nucleic Acids Research, 43, pp. doi: 10.1093/nar/gkv007</p>
<p><img
        class="lazyload"
        src="/MajorbioHuGo/svg/loading.min.svg"
        data-src="limma.jpg"
        data-srcset="/MajorbioHuGo/posts/scrnaseq-batch-correction/limma.jpg, limma.jpg 1.5x, /MajorbioHuGo/posts/scrnaseq-batch-correction/limma.jpg 2x"
        data-sizes="auto"
        alt="/MajorbioHuGo/posts/scrnaseq-batch-correction/limma.jpg"
        title="limma.jpg" /></p>
<h2 id="harmony">harmony</h2>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="c1">#!/usr/bin/env Rscript3.6.1</span>
<span class="nf">suppressMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;getopt&#34;</span><span class="p">))</span>
<span class="nf">options</span><span class="p">(</span><span class="n">bitmapType</span><span class="o">=</span><span class="s">&#39;cairo&#39;</span><span class="p">)</span>
<span class="n">spec</span> <span class="o">=</span> <span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span>
    <span class="s">&#39;rds&#39;</span><span class="p">,</span><span class="s">&#39;r&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
    <span class="s">&#39;outdir&#39;</span><span class="p">,</span><span class="s">&#39;o&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
    <span class="s">&#39;help&#39;</span><span class="p">,</span><span class="s">&#39;h&#39;</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="s">&#39;logical&#39;</span>
    <span class="p">),</span> <span class="n">byrow</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="m">4</span><span class="p">);</span>
<span class="n">opt</span> <span class="o">=</span> <span class="nf">getopt</span><span class="p">(</span><span class="n">spec</span><span class="p">);</span>
<span class="n">print_usage</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">spec</span><span class="o">=</span><span class="kc">NULL</span><span class="p">){</span>
  <span class="nf">cat</span><span class="p">(</span><span class="nf">getopt</span><span class="p">(</span><span class="n">spec</span><span class="p">,</span> <span class="n">usage</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">));</span>
  <span class="nf">cat</span><span class="p">(</span><span class="s">&#34;Usage example: \n&#34;</span><span class="p">);</span>
  <span class="nf">cat</span><span class="p">(</span><span class="s">&#34;Usage example:
</span><span class="s">    Rscript harmony.r --i  --o
</span><span class="s">    Usage:
</span><span class="s">    --rds       input file rds
</span><span class="s">    --outdir    outdir
</span><span class="s">    --help              usage\n&#34;</span>
    <span class="p">);</span>
  <span class="nf">q</span><span class="p">(</span><span class="n">status</span><span class="o">=</span><span class="m">1</span><span class="p">);</span>
<span class="p">}</span>

<span class="nf">if </span><span class="p">(</span> <span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">help</span><span class="p">))</span> <span class="p">{</span><span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">rds</span><span class="p">)){</span><span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">)){</span><span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)}</span>

<span class="nf">suppressMessages</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;harmony&#34;</span><span class="p">),</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;magrittr&#34;</span><span class="p">),</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;Seurat&#34;</span><span class="p">),</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;dplyr&#34;</span><span class="p">)))</span>
<span class="n">scRNA</span> <span class="o">&lt;-</span> <span class="nf">readRDS</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">rds</span><span class="p">)</span>
<span class="c1"># Normalize/find features</span>
<span class="n">scRNA</span> <span class="o">&lt;-</span> <span class="nf">NormalizeData</span><span class="p">(</span><span class="n">scRNA</span><span class="p">)</span>
<span class="n">scRNA</span> <span class="o">&lt;-</span> <span class="nf">FindVariableFeatures</span><span class="p">(</span><span class="n">scRNA</span><span class="p">,</span> <span class="n">selection.method</span> <span class="o">=</span> <span class="s">&#34;vst&#34;</span><span class="p">,</span> <span class="n">nfeatures</span> <span class="o">=</span> <span class="m">2000</span><span class="p">)</span>
<span class="n">scRNA</span> <span class="o">&lt;-</span> <span class="nf">ScaleData</span><span class="p">(</span><span class="n">scRNA</span><span class="p">,</span><span class="n">vars.to.regress</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&#34;nUMI&#34;</span><span class="p">,</span> <span class="s">&#34;percent.mito&#34;</span><span class="p">,</span><span class="s">&#34;percent.ribo&#34;</span><span class="p">,</span><span class="s">&#34;sample&#34;</span><span class="p">))</span>
<span class="n">scRNA</span> <span class="o">&lt;-</span> <span class="nf">RunPCA</span><span class="p">(</span><span class="n">scRNA</span><span class="p">,</span> <span class="n">features</span> <span class="o">=</span> <span class="nf">VariableFeatures</span><span class="p">(</span><span class="n">object</span> <span class="o">=</span> <span class="n">scRNA</span><span class="p">),</span><span class="n">verbose</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>
<span class="c1"># batch</span>
<span class="n">scRNA</span> <span class="o">&lt;-</span> <span class="n">scRNA</span> <span class="o">%&gt;%</span> <span class="nf">RunHarmony</span><span class="p">(</span><span class="s">&#34;orig.ident&#34;</span><span class="p">,</span> <span class="n">plot_convergence</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">,</span><span class="s">&#34;harmony.RData&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#34;&#34;</span><span class="p">)</span>
</code></pre></div><p><img
        class="lazyload"
        src="/MajorbioHuGo/svg/loading.min.svg"
        data-src="harmony_batch.jpg"
        data-srcset="/MajorbioHuGo/posts/scrnaseq-batch-correction/harmony_batch.jpg, harmony_batch.jpg 1.5x, /MajorbioHuGo/posts/scrnaseq-batch-correction/harmony_batch.jpg 2x"
        data-sizes="auto"
        alt="/MajorbioHuGo/posts/scrnaseq-batch-correction/harmony_batch.jpg"
        title="harmony_batch.jpg" /></p>
<p>参考文献：Korsunsky I, Millard N, Fan J, et al. <a href="https://www.nature.com/articles/s41592-019-0619-0" target="_blank" rel="noopener noreffer">Fast, sensitive and accurate integration of single-cell data with Harmony</a>. Nature methods, 2019: 1-8.</p>
<p><img
        class="lazyload"
        src="/MajorbioHuGo/svg/loading.min.svg"
        data-src="harmony.jpg"
        data-srcset="/MajorbioHuGo/posts/scrnaseq-batch-correction/harmony.jpg, harmony.jpg 1.5x, /MajorbioHuGo/posts/scrnaseq-batch-correction/harmony.jpg 2x"
        data-sizes="auto"
        alt="/MajorbioHuGo/posts/scrnaseq-batch-correction/harmony.jpg"
        title="harmony.jpg" /></p>
<h2 id="liger">LIGER</h2>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="c1">#!/usr/bin/env Rscript3.6.1</span>
<span class="nf">suppressMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;getopt&#34;</span><span class="p">))</span>
<span class="nf">options</span><span class="p">(</span><span class="n">bitmapType</span><span class="o">=</span><span class="s">&#39;cairo&#39;</span><span class="p">)</span>
<span class="n">spec</span> <span class="o">=</span> <span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span>
    <span class="s">&#39;rds&#39;</span><span class="p">,</span><span class="s">&#39;r&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
    <span class="s">&#39;outdir&#39;</span><span class="p">,</span><span class="s">&#39;o&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
    <span class="s">&#39;help&#39;</span><span class="p">,</span><span class="s">&#39;h&#39;</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="s">&#39;logical&#39;</span>
    <span class="p">),</span> <span class="n">byrow</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="m">4</span><span class="p">);</span>
<span class="n">opt</span> <span class="o">=</span> <span class="nf">getopt</span><span class="p">(</span><span class="n">spec</span><span class="p">);</span>
<span class="n">print_usage</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">spec</span><span class="o">=</span><span class="kc">NULL</span><span class="p">){</span>
  <span class="nf">cat</span><span class="p">(</span><span class="nf">getopt</span><span class="p">(</span><span class="n">spec</span><span class="p">,</span> <span class="n">usage</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">));</span>
  <span class="nf">cat</span><span class="p">(</span><span class="s">&#34;Usage example: \n&#34;</span><span class="p">);</span>
  <span class="nf">cat</span><span class="p">(</span><span class="s">&#34;Usage example:
</span><span class="s">    Rscript liger.r --i  --o
</span><span class="s">    Usage:
</span><span class="s">    --rds	input file rds
</span><span class="s">    --outdir	outdir
</span><span class="s">    --help		usage\n&#34;</span>
    <span class="p">);</span>
  <span class="nf">q</span><span class="p">(</span><span class="n">status</span><span class="o">=</span><span class="m">1</span><span class="p">);</span>
<span class="p">}</span>

<span class="nf">if </span><span class="p">(</span> <span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">help</span><span class="p">))</span> <span class="p">{</span><span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">rds</span><span class="p">)){</span><span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">)){</span><span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)}</span>

<span class="nf">suppressMessages</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;liger&#34;</span><span class="p">),</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;Seurat&#34;</span><span class="p">)))</span>

<span class="n">scRNA</span> <span class="o">&lt;-</span> <span class="nf">readRDS</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">rds</span><span class="p">)</span>
<span class="n">x2</span> <span class="o">&lt;-</span> <span class="n">scRNA</span><span class="o">@</span><span class="n">assays</span><span class="o">$</span><span class="n">RNA</span><span class="o">@</span><span class="n">counts[</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">7205</span><span class="n">]</span>
<span class="n">x4</span> <span class="o">&lt;-</span> <span class="n">scRNA</span><span class="o">@</span><span class="n">assays</span><span class="o">$</span><span class="n">RNA</span><span class="o">@</span><span class="n">counts[</span><span class="p">,</span><span class="m">7206</span><span class="o">:</span><span class="m">16309</span><span class="n">]</span>
<span class="n">ifnb_liger</span> <span class="o">&lt;-</span> <span class="nf">createLiger</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">x2</span> <span class="o">=</span> <span class="n">x2</span><span class="p">,</span> <span class="n">x4</span> <span class="o">=</span> <span class="n">x4</span><span class="p">))</span>
<span class="n">ifnb_liger</span> <span class="o">&lt;-</span> <span class="nf">normalize</span><span class="p">(</span><span class="n">ifnb_liger</span><span class="p">)</span>
<span class="n">ifnb_liger</span> <span class="o">&lt;-</span> <span class="nf">selectGenes</span><span class="p">(</span><span class="n">ifnb_liger</span><span class="p">)</span>
<span class="n">ifnb_liger</span> <span class="o">&lt;-</span> <span class="nf">scaleNotCenter</span><span class="p">(</span><span class="n">ifnb_liger</span><span class="p">)</span>
<span class="n">ifnb_liger</span> <span class="o">&lt;-</span> <span class="nf">optimizeALS</span><span class="p">(</span><span class="n">ifnb_liger</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="m">20</span><span class="p">)</span>
<span class="n">ifnb_liger</span> <span class="o">&lt;-</span> <span class="nf">quantile_norm</span><span class="p">(</span><span class="n">ifnb_liger</span><span class="p">)</span>
<span class="nf">save.image</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">,</span><span class="s">&#34;LIGER_result.RData&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#34;&#34;</span><span class="p">))</span>
<span class="nf">proc.time</span><span class="p">()</span>
</code></pre></div><p>使用liger去批次后所得矩阵，计算差异表达基因。由于liger去批次后数据矩阵结果0值过多，会导致计算差异表达时数据被删除的问题。</p>
<p><img
        class="lazyload"
        src="/MajorbioHuGo/svg/loading.min.svg"
        data-src="liger_DE.jpg"
        data-srcset="/MajorbioHuGo/posts/scrnaseq-batch-correction/liger_DE.jpg, liger_DE.jpg 1.5x, /MajorbioHuGo/posts/scrnaseq-batch-correction/liger_DE.jpg 2x"
        data-sizes="auto"
        alt="/MajorbioHuGo/posts/scrnaseq-batch-correction/liger_DE.jpg"
        title="liger_DE.jpg" /></p>
<p>参考文献：Welch J, Kozareva V, Ferreira A, et al. Integrative inference of brain cell similarities and differences from single-cell genomics[J]. bioRxiv, 2018: 459891.</p>
<p><img
        class="lazyload"
        src="/MajorbioHuGo/svg/loading.min.svg"
        data-src="LIGER.jpg"
        data-srcset="/MajorbioHuGo/posts/scrnaseq-batch-correction/LIGER.jpg, LIGER.jpg 1.5x, /MajorbioHuGo/posts/scrnaseq-batch-correction/LIGER.jpg 2x"
        data-sizes="auto"
        alt="/MajorbioHuGo/posts/scrnaseq-batch-correction/LIGER.jpg"
        title="LIGER.jpg" /></p>
<h2 id="mnn-correct">MNN Correct</h2>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="c1">#!/usr/bin/env Rscript3.6.1</span>
<span class="nf">suppressMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;getopt&#34;</span><span class="p">))</span>
<span class="nf">options</span><span class="p">(</span><span class="n">bitmapType</span><span class="o">=</span><span class="s">&#39;cairo&#39;</span><span class="p">)</span>
<span class="n">spec</span> <span class="o">=</span> <span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span>
    <span class="s">&#39;rds&#39;</span><span class="p">,</span><span class="s">&#39;r&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
    <span class="s">&#39;outdir&#39;</span><span class="p">,</span><span class="s">&#39;o&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
    <span class="s">&#39;help&#39;</span><span class="p">,</span><span class="s">&#39;h&#39;</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="s">&#39;logical&#39;</span>
    <span class="p">),</span> <span class="n">byrow</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="m">4</span><span class="p">);</span>
<span class="n">opt</span> <span class="o">=</span> <span class="nf">getopt</span><span class="p">(</span><span class="n">spec</span><span class="p">);</span>
<span class="n">print_usage</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">spec</span><span class="o">=</span><span class="kc">NULL</span><span class="p">){</span>
  <span class="nf">cat</span><span class="p">(</span><span class="nf">getopt</span><span class="p">(</span><span class="n">spec</span><span class="p">,</span> <span class="n">usage</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">));</span>
  <span class="nf">cat</span><span class="p">(</span><span class="s">&#34;Usage example: \n&#34;</span><span class="p">);</span>
  <span class="nf">cat</span><span class="p">(</span><span class="s">&#34;Usage example:
</span><span class="s">    Rscript mnncorrect.r --i  --o
</span><span class="s">    Usage:
</span><span class="s">    --rds       input file rds
</span><span class="s">    --outdir    outdir
</span><span class="s">    --help              usage\n&#34;</span>
    <span class="p">);</span>
  <span class="nf">q</span><span class="p">(</span><span class="n">status</span><span class="o">=</span><span class="m">1</span><span class="p">);</span>
<span class="p">}</span>

<span class="nf">if </span><span class="p">(</span> <span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">help</span><span class="p">))</span> <span class="p">{</span><span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">rds</span><span class="p">)){</span><span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">)){</span><span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)}</span>


<span class="nf">suppressMessages</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;Seurat&#34;</span><span class="p">),</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;batchelor&#34;</span><span class="p">)))</span>

<span class="n">scRNA</span> <span class="o">&lt;-</span> <span class="nf">readRDS</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">rds</span><span class="p">)</span>
<span class="n">test1</span> <span class="o">&lt;-</span> <span class="n">scRNA</span><span class="o">@</span><span class="n">assays</span><span class="o">$</span><span class="n">RNA</span><span class="o">@</span><span class="n">counts[</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">7205</span><span class="n">]</span>
<span class="n">test2</span> <span class="o">&lt;-</span> <span class="n">scRNA</span><span class="o">@</span><span class="n">assays</span><span class="o">$</span><span class="n">RNA</span><span class="o">@</span><span class="n">counts[</span><span class="p">,</span><span class="m">7206</span><span class="o">:</span><span class="m">16309</span><span class="n">]</span>
<span class="n">test1</span> <span class="o">&lt;-</span> <span class="nf">as.matrix</span><span class="p">(</span><span class="n">test1</span><span class="p">)</span>
<span class="n">test2</span> <span class="o">&lt;-</span> <span class="nf">as.matrix</span><span class="p">(</span><span class="n">test2</span><span class="p">)</span>
<span class="n">out</span> <span class="o">&lt;-</span> <span class="nf">mnnCorrect</span><span class="p">(</span><span class="n">test1</span><span class="p">,</span><span class="n">test2</span><span class="p">)</span>
<span class="nf">save.image</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">,</span><span class="s">&#34;mnnCorrect.RData&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#34;&#34;</span><span class="p">))</span>
<span class="nf">proc.time</span><span class="p">()</span>
</code></pre></div><p><img
        class="lazyload"
        src="/MajorbioHuGo/svg/loading.min.svg"
        data-src="MNNCorrect_DE.jpg"
        data-srcset="/MajorbioHuGo/posts/scrnaseq-batch-correction/MNNCorrect_DE.jpg, MNNCorrect_DE.jpg 1.5x, /MajorbioHuGo/posts/scrnaseq-batch-correction/MNNCorrect_DE.jpg 2x"
        data-sizes="auto"
        alt="/MajorbioHuGo/posts/scrnaseq-batch-correction/MNNCorrect_DE.jpg"
        title="MNNCorrect_DE.jpg" /></p>
<p>参考文献：Haghverdi L, Lun A T L, Morgan M D, et al. Batch effects in single-cell RNA-sequencing data are corrected by matching mutual nearest neighbors[J]. Nature biotechnology, 2018, 36(5): 421-427.</p>
<p><img
        class="lazyload"
        src="/MajorbioHuGo/svg/loading.min.svg"
        data-src="bacthelor.jpg"
        data-srcset="/MajorbioHuGo/posts/scrnaseq-batch-correction/bacthelor.jpg, bacthelor.jpg 1.5x, /MajorbioHuGo/posts/scrnaseq-batch-correction/bacthelor.jpg 2x"
        data-sizes="auto"
        alt="/MajorbioHuGo/posts/scrnaseq-batch-correction/bacthelor.jpg"
        title="MNNCorrect.jpg" /></p>
<h2 id="fastmnn">fastMNN</h2>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="c1">#!/usr/bin/env Rscript3.6.1</span>
<span class="nf">suppressMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;getopt&#34;</span><span class="p">))</span>
<span class="nf">options</span><span class="p">(</span><span class="n">bitmapType</span><span class="o">=</span><span class="s">&#39;cairo&#39;</span><span class="p">)</span>
<span class="n">spec</span> <span class="o">=</span> <span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span>
    <span class="s">&#39;rds&#39;</span><span class="p">,</span><span class="s">&#39;r&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
    <span class="s">&#39;outdir&#39;</span><span class="p">,</span><span class="s">&#39;o&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
    <span class="s">&#39;help&#39;</span><span class="p">,</span><span class="s">&#39;h&#39;</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="s">&#39;logical&#39;</span>
    <span class="p">),</span> <span class="n">byrow</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="m">4</span><span class="p">);</span>
<span class="n">opt</span> <span class="o">=</span> <span class="nf">getopt</span><span class="p">(</span><span class="n">spec</span><span class="p">);</span>
<span class="n">print_usage</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">spec</span><span class="o">=</span><span class="kc">NULL</span><span class="p">){</span>
  <span class="nf">cat</span><span class="p">(</span><span class="nf">getopt</span><span class="p">(</span><span class="n">spec</span><span class="p">,</span> <span class="n">usage</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">));</span>
  <span class="nf">cat</span><span class="p">(</span><span class="s">&#34;Usage example: \n&#34;</span><span class="p">);</span>
  <span class="nf">cat</span><span class="p">(</span><span class="s">&#34;Usage example:
</span><span class="s">    Rscript fastmnn.r --i  --o
</span><span class="s">    Usage:
</span><span class="s">    --rds       input file rds
</span><span class="s">    --outdir    outdir
</span><span class="s">    --help              usage\n&#34;</span>
    <span class="p">);</span>
  <span class="nf">q</span><span class="p">(</span><span class="n">status</span><span class="o">=</span><span class="m">1</span><span class="p">);</span>
<span class="p">}</span>

<span class="nf">if </span><span class="p">(</span> <span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">help</span><span class="p">))</span> <span class="p">{</span><span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">rds</span><span class="p">)){</span><span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">)){</span><span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)}</span>


<span class="nf">suppressMessages</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;Seurat&#34;</span><span class="p">),</span><span class="nf">library</span><span class="p">(</span><span class="s">&#34;batchelor&#34;</span><span class="p">)))</span>

<span class="n">scRNA</span> <span class="o">&lt;-</span> <span class="nf">readRDS</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">rds</span><span class="p">)</span>
<span class="n">test1</span> <span class="o">&lt;-</span> <span class="n">scRNA</span><span class="o">@</span><span class="n">assays</span><span class="o">$</span><span class="n">RNA</span><span class="o">@</span><span class="n">counts[</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">7205</span><span class="n">]</span>
<span class="n">test2</span> <span class="o">&lt;-</span> <span class="n">scRNA</span><span class="o">@</span><span class="n">assays</span><span class="o">$</span><span class="n">RNA</span><span class="o">@</span><span class="n">counts[</span><span class="p">,</span><span class="m">7206</span><span class="o">:</span><span class="m">16309</span><span class="n">]</span>
<span class="n">test1</span> <span class="o">&lt;-</span> <span class="nf">as.matrix</span><span class="p">(</span><span class="n">test1</span><span class="p">)</span>
<span class="n">test2</span> <span class="o">&lt;-</span> <span class="nf">as.matrix</span><span class="p">(</span><span class="n">test2</span><span class="p">)</span>
<span class="n">out</span> <span class="o">&lt;-</span> <span class="nf">fastMNN</span><span class="p">(</span><span class="n">test1</span><span class="p">,</span><span class="n">test2</span><span class="p">)</span>
<span class="nf">save.image</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">,</span><span class="s">&#34;mnnCorrect.RData&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#34;&#34;</span><span class="p">))</span>
<span class="nf">proc.time</span><span class="p">()</span>
</code></pre></div><h1 id="参考文献">参考文献</h1>
<p>Haghverdi L, Lun A T L, Morgan M D, et al. <a href="https://www.nature.com/articles/nbt.4091" target="_blank" rel="noopener noreffer">Batch effects in single-cell RNA-sequencing data are corrected by matching mutual nearest neighbors</a>. Nature biotechnology, 2018, 36(5): 421-427.</p>
<p>Tran, H.T.N., Ang, K.S., Chevrier, M. et al. <a href="https://doi.org/10.1186/s13059-019-1850-9" target="_blank" rel="noopener noreffer">A benchmark of batch-effect correction methods for single-cell RNA sequencing data</a>. Genome Biol 21, 12 (2020).</p>
</div><div class="post-footer" id="post-footer">
    <div class="post-info">
        <div class="post-info-line">
            <div class="post-info-mod">
                <span>更新于 2020-12-15</span>
            </div>
            <div class="post-info-license"></div>
        </div>
        <div class="post-info-line">
            <div class="post-info-md"></div>
            <div class="post-info-share">
                <span></span>
            </div>
        </div>
    </div>

    <div class="post-info-more">
        <section class="post-tags"><i class="fas fa-tags fa-fw"></i>&nbsp;<a href="/MajorbioHuGo/tags/%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94/">批次效应</a>,&nbsp;<a href="/MajorbioHuGo/tags/scrnaseq/">scRNAseq</a>,&nbsp;<a href="/MajorbioHuGo/tags/rnaseq/">RNAseq</a>,&nbsp;<a href="/MajorbioHuGo/tags/%E5%8D%95%E7%BB%86%E8%83%9E/">单细胞</a></section>
        <section>
            <span><a href="javascript:void(0);" onclick="window.history.back();">返回</a></span>&nbsp;|&nbsp;<span><a href="/MajorbioHuGo/">主页</a></span>
        </section>
    </div>

    <div class="post-nav"><a href="/MajorbioHuGo/posts/mapping-make-fake-data/" class="prev" rel="prev" title="根据比对率生产假的原始序列数据"><i class="fas fa-angle-left fa-fw"></i>根据比对率生产假的原始序列数据</a>
            <a href="/MajorbioHuGo/posts/probabilistic-estimation-of-expression-residuals/" class="next" rel="next" title="Using probabilistic estimation of expression residuals (PEER) to obtain increased power and interpretability of gene expression analyses">Using probabilistic estimation of expression residuals (PEER) to obtain increased power and interpretability of gene expression analyses<i class="fas fa-angle-right fa-fw"></i></a></div>
</div>
</article></div>
            </main><footer class="footer">
        <div class="footer-container"><div class="footer-line">由 <a href="https://gohugo.io/" target="_blank" rel="noopener noreffer" title="Hugo 0.79.0">Hugo</a> 强力驱动 | 主题 - <a href="https://github.com/dillonzq/LoveIt" target="_blank" rel="noopener noreffer" title="LoveIt 0.2.10"><i class="far fa-kiss-wink-heart fa-fw"></i> LoveIt</a>
                </div><div class="footer-line"><i class="far fa-copyright fa-fw"></i><span itemprop="copyrightYear">2020</span><span class="author" itemprop="copyrightHolder">&nbsp;<a href="/MajorbioHuGo/" target="_blank"></a></span></div>
        </div>
    </footer></div>

        <div id="fixed-buttons"><a href="#" id="back-to-top" class="fixed-button" title="回到顶部">
                <i class="fas fa-arrow-up fa-fw"></i>
            </a><a href="#" id="view-comments" class="fixed-button" title="查看评论">
                <i class="fas fa-comment fa-fw"></i>
            </a>
        </div><link rel="stylesheet" href="/MajorbioHuGo/lib/lightgallery/lightgallery.min.css"><script type="text/javascript" src="/MajorbioHuGo/lib/smooth-scroll/smooth-scroll.min.js"></script><script type="text/javascript" src="/MajorbioHuGo/lib/autocomplete/autocomplete.min.js"></script><script type="text/javascript" src="/MajorbioHuGo/lib/algoliasearch/algoliasearch-lite.umd.min.js"></script><script type="text/javascript" src="/MajorbioHuGo/lib/lazysizes/lazysizes.min.js"></script><script type="text/javascript" src="/MajorbioHuGo/lib/lightgallery/lightgallery.min.js"></script><script type="text/javascript" src="/MajorbioHuGo/lib/lightgallery/lg-thumbnail.min.js"></script><script type="text/javascript" src="/MajorbioHuGo/lib/lightgallery/lg-zoom.min.js"></script><script type="text/javascript" src="/MajorbioHuGo/lib/clipboard/clipboard.min.js"></script><script type="text/javascript">window.config={"code":{"copyTitle":"复制到剪贴板","maxShownLines":10},"comment":{},"lightGallery":{"actualSize":false,"exThumbImage":"data-thumbnail","hideBarsDelay":2000,"selector":".lightgallery","speed":400,"thumbContHeight":80,"thumbWidth":80,"thumbnail":true},"search":{"algoliaAppID":"VX4BXOJBC5","algoliaIndex":"majorbio.zh-cn","algoliaSearchKey":"53790f167edb548b217bfe91dd891ac0","highlightTag":"em","maxResultLength":10,"noResultsFound":"没有找到结果","snippetLength":50,"type":"algolia"}};</script><script type="text/javascript" src="/MajorbioHuGo/js/theme.min.js"></script></body>
</html>
